### DISCUSSION TITLE
Define intial grain structure for grain growth model

### DISCUSSION THREAD

**Discussion by jackson911007** [idaholab/moose#26003] [Q&A General]
Hi, 

I am new to phase-field modeling and find Moose extremely helpful and easy to implement. I tried the grain growth model and successfully ran it at different times and temperatures. Now I am moving on to defining my own initial grain structure using a txt file that describes a distribution of grain size. Could you kindly point me to where I can do this in the .i file or an example?

Thanks for your help, 

```
(Here is the code I use)
# This simulation predicts GB migration of a 2D copper polycrystal with 100 grains represented with 8 order parameters
# Mesh adaptivity and time step adaptivity are used
# An AuxVariable is used to calculate the grain boundary locations
# Postprocessors are used to record time step and the number of grains

[Mesh]
  # Mesh block.  Meshes can be read in or automatically generated
  type = GeneratedMesh
  dim = 2 # Problem dimension
  nx = 44 # Number of elements in the x-direction
  ny = 44 # Number of elements in the y-direction
  xmax = 1000 # maximum x-coordinate of the mesh
  ymax = 1000 # maximum y-coordinate of the mesh
  elem_type = QUAD4 # Type of elements used in the mesh
  uniform_refine = 2 # Initial uniform refinement of the mesh
[]

[GlobalParams]
  # Parameters used by several kernels that are defined globally to simplify input file
  op_num = 8 # Number of order parameters used
  var_name_base = gr # Base name of grains
[]

[Modules]
  [PhaseField]
    [GrainGrowth]
    []
  []
[]

[UserObjects]
  [voronoi]
    type = PolycrystalVoronoi
    grain_num = 100 # Number of grains
    rand_seed = 10
    int_width = 7
  []
  [grain_tracker]
    type = GrainTracker
  []
[]

[ICs]
  [PolycrystalICs]
    [PolycrystalColoringIC]
      polycrystal_ic_uo = voronoi
    []
  []
[]

[AuxVariables]
  # Dependent variables
  [unique_grains]
    order = CONSTANT
    family = MONOMIAL
  []
  [var_indices]
    order = CONSTANT
    family = MONOMIAL
  []
[]

[AuxKernels]
  # AuxKernel block, defining the equations used to calculate the auxvars
  [bnds_aux]
    # AuxKernel that calculates the GB term
    type = BndsCalcAux
    variable = bnds
    execute_on = 'initial timestep_end'
  []
  [unique_grains]
    type = FeatureFloodCountAux
    variable = unique_grains
    flood_counter = grain_tracker
    field_display = UNIQUE_REGION
    execute_on = 'initial timestep_end'
  []
  [var_indices]
    type = FeatureFloodCountAux
    variable = var_indices
    flood_counter = grain_tracker
    field_display = VARIABLE_COLORING
    execute_on = 'initial timestep_end'
  []
[]

[BCs]
  # Boundary Condition block
  [Periodic]
    [All]
      auto_direction = 'x y' # Makes problem periodic in the x and y directions
    []
  []
[]

[Materials]
  [CuGrGr]
    # Material properties
    type = GBEvolution
    T = 450 # Constant temperature of the simulation (for mobility calculation)
    wGB = 14 # Width of the diffuse GB
    GBmob0 = 2.5e-6 #m^4(Js) for copper from Schoenfelder1997
    Q = 0.23 #eV for copper from Schoenfelder1997
    GBenergy = 0.708 #J/m^2 from Schoenfelder1997
  []
[]

[Postprocessors]
  # Scalar postprocessors
  [dt]
    # Outputs the current time step
    type = TimestepSize
  []
[]

[Executioner]
  type = Transient # Type of executioner, here it is transient with an adaptive time step
  scheme = bdf2 # Type of time integration (2nd order backward euler), defaults to 1st order backward euler

  #Preconditioned JFNK (default)
  solve_type = 'PJFNK'

  # Uses newton iteration to solve the problem.
  petsc_options_iname = '-pc_type -pc_hypre_type'
  petsc_options_value = 'hypre boomeramg'

  l_max_its = 50 # Max number of linear iterations
  l_tol = 1e-4 # Relative tolerance for linear solves
  nl_max_its = 10 # Max number of nonlinear iterations

  end_time = 4000

  [TimeStepper]
    type = IterationAdaptiveDT
    dt = 20 # Initial time step.  In this simulation it changes.
    optimal_iterations = 6 # Time step will adapt to maintain this number of nonlinear iterations
  []

  [Adaptivity]
    # Block that turns on mesh adaptivity. Note that mesh will never coarsen beyond initial mesh (before uniform refinement)
    initial_adaptivity = 2 # Number of times mesh is adapted to initial condition
    refine_fraction = 0.8 # Fraction of high error that will be refined
    coarsen_fraction = 0.05 # Fraction of low error that will coarsened
    max_h_level = 2 # Max number of refinements used, starting from initial mesh (before uniform refinement)
  []
[]

[Outputs]
  exodus = true # Exodus file will be outputted
  csv = true
[]
```

**Comment by permcody** [comment]
Hi @jackson911007,

There are a few different options you can take a look at here to help assist in creating your own custom initial conditions: 
- EBSD Reader - This is a standardize format for supplying the grain information of each cell in a regularized grid. This gives you a ton of flexibility to shape your initial grain structure however you please. The only downside to this is that you'll likely need to write a little script to compute this information if you are starting from higher level information like grain centers or as you said distributions of sizes. [PolycrystalEBSD](https://mooseframework.inl.gov/source/userobjects/PolycrystalEBSD.html) In particular, take a look at a sample EBSD file: [Sample EBSD File](https://github.com/idaholab/moose/blob/next/modules/phase_field/test/tests/grain_tracker_test/ebsd_9.txt)
- Voronoi - While this won't exactly meet your needs, I would recommend starting with this object because it's conceptually simple and once you get the hang of what it's doing, you can probably extend it to meet your needs. All it does is generate random points in a given domain and then MOOSE computes the Voronoi tessellation of those points. [PolycrystalVoronoi](https://mooseframework.inl.gov/source/userobjects/PolycrystalVoronoi.html)
- Roll your own Initial Condition - After looking at the above two examples, you can build your own IC. We've worked hard to create the simpliest interface possible, you simply need to create C++ code that can return the dominate grain for every position in your mesh. That's it. That method gets called repeatedly to setup the initial condition with MOOSE doing all the heavy lifting of creating the smooth interfaces (once you start the simulation) and applies coloring algorithms so you can use reduced order parameter modeling. [PolycrystalICs](https://mooseframework.inl.gov/modules/phase_field/ICs/PolycrystalICs.html)

**Comment by jackson911007** [comment]
Thanks a lot for the information. I will try it out.

Best regards,

Jackson

**Comment by jackson911007** [comment]
Hi @permcody, 

thanks for your recommendation. I decide to go through the second method first. I tried several methods for polycrystal and found  Voronoi might be a better fit. However, I want increase the size distribution of it. Is it possible to do through this script?

Thanks a lot, 

Jackson

**Reply by GiudGiud** [reply]
So you want bigger voronoi grains?
For the same mesh, you need to have less grains then
This can be set in the parameters of the PolycrystalVoronoi object with grain_num 
or try working on the grain patch size instead

**Comment by jackson911007** [comment]
Thanks for the tip! Sorry I do not explain my question correctly. The
question I have is that could I adjust the standard deviation of the
distribution (make it spread broader)?

**Reply by GiudGiud** [reply]
Ah good question. I dont know if we have this capability, I dont see it in the parameters. 
This could probably be added in the code for the generation of the points underlying the voronoi partition

**Comment by jackson911007** [comment]
Thanks, we talking about the partition. Does it mean I  will adjust the
number grain or the nx, ny parameter in the mesh to increase the std?

**Reply by GiudGiud** [reply]
increasing nx / ny will make the mesh bigger but the distribution of the sizes of the grains will barely change

the number of grain will certainly change the average grain size, but the std deviation of the sizes may not change

**Comment by jackson911007** [comment]
Got it. Thanks, do you suggest changing any other parameter in the .i file
to increase the std?

**Reply by GiudGiud** [reply]
I dont see an option in the parameters. You will have to check the code

**Comment by jackson911007** [comment]
Ok, I'll check. Thanks for the suggestion

**Comment by permcody** [comment]
I believe you can always supply your own voronoi grain centers if you wish.
Then can exactly control your grain structure. If that option isn’t
currently available, it would be trivial to expose it to the input file.

**Reply by jackson911007** [reply]
Hi Permcody, 

I still have a question so how could I "supply your own voronoi grain centers"?

**Reply by GiudGiud** [reply]
I think you can use the file_name parameter and specify a CSV file, in the PolycrystalVoronoi

https://mooseframework.inl.gov/source/userobjects/PolycrystalVoronoi.html

**Reply by jackson911007** [reply]
@GiudGiud Thanks for your recommendation. Is there any example to define this CSV? Thanks,

**Reply by GiudGiud** [reply]
you generate it with external software. I would just use python and implement whichever distribution law you wanted

**Reply by jackson911007** [reply]
This seems very helpful. Sorry I am very new in Phase field. Maybe this is a damp question. What I have now is a list of lognormal distribution sample size s.g. samples = rng.lognormal(mean=np.log(135), sigma=1.14, size=200). Do you have any suggestion how I can generate a csv file like this? Thanks,

**Reply by GiudGiud** [reply]
Python or matlab will both be able to generate samples from a log normal distribution. Then you can output them to csv using the `csv` library in python for example.

**Reply by jackson911007** [reply]
Just to confirm the .csv file  is a list of grain size? Above, the samples I generated is a list of 200 samples in a numpy array: np.array([527.79341127,   62.9196671 ,   34.80595431,  310.05654114, 55.32453322,  121.93137714,  156.98187213,  674.51899689, 86.56465744...]). For the csv file, does it mean I only need to save this array in a .csv file and that's all?

**Reply by GiudGiud** [reply]
I think it s a csv file of centroids of the grains, not sizes

**Reply by jackson911007** [reply]
Hello Giud, sorry I might not totally get. Could I ask for your further assistance on this? I sent out a PM to you at gmail.

**Comment by jackson911007** [comment]
Thank you for your response. I will give it a try.
